presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -0,0 +1,424 @@
+        if (multiJoinNode.getSources().size() < 2) {
+            // Possible when e.g. joinNode has non-deterministic filter. Reordering is generally not allowed in this case.
+            return Result.empty();
+        }
+
+        JoinEnumerator joinEnumerator = new JoinEnumerator(
+                context.getSession(),
+                context.getCostProvider(),
+                costComparator,
+                context.getIdAllocator(),
+                multiJoinNode.getFilter(),
+                context.getLookup());
+        JoinEnumerationResult result = joinEnumerator.chooseJoinOrder(multiJoinNode.getSources(), multiJoinNode.getOutputSymbols());
+        if (!result.getPlanNode().isPresent()) {
+            return Result.empty();
+        }
+        if (result.getCost().hasUnknownComponents() || result.getCost().equals(INFINITE_COST)) {
+            return Result.empty();
+        }
+        return Result.ofPlanNode(result.getPlanNode().get());
+    }
+
+    @VisibleForTesting
+    static class JoinEnumerator
+    {
+        private final Session session;
+        private final CostProvider costProvider;
+        // Using Ordering to facilitate rule determinism
+        private final Ordering<JoinEnumerationResult> resultComparator;
+        private final PlanNodeIdAllocator idAllocator;
+        private final EqualityInference allInference;
+        private final Expression allFilter;
+        private final Lookup lookup;
+
+        private final Map<Set<PlanNode>, JoinEnumerationResult> memo = new HashMap<>();
+
+        @VisibleForTesting
+        JoinEnumerator(Session session, CostProvider costProvider, CostComparator costComparator, PlanNodeIdAllocator idAllocator, Expression filter, Lookup lookup)
+        {
+            this.session = requireNonNull(session, "session is null");
+            this.costProvider = requireNonNull(costProvider, "costProvider is null");
+            this.resultComparator = costComparator.forSession(session).onResultOf(result -> result.cost);
+            this.idAllocator = requireNonNull(idAllocator, "idAllocator is null");
+            this.allInference = createEqualityInference(filter);
+            this.allFilter = requireNonNull(filter, "filter is null");
+            this.lookup = requireNonNull(lookup, "lookup is null");
+        }
+
+        private JoinEnumerationResult chooseJoinOrder(List<PlanNode> sources, List<Symbol> outputSymbols)
+        {
+            Set<PlanNode> multiJoinKey = ImmutableSet.copyOf(sources);
+            JoinEnumerationResult bestResult = memo.get(multiJoinKey);
+            if (bestResult == null) {
+                checkState(sources.size() > 1, "sources size is less than or equal to one");
+                ImmutableList.Builder<JoinEnumerationResult> resultBuilder = ImmutableList.builder();
+                Set<Set<Integer>> partitions = generatePartitions(sources.size()).collect(toImmutableSet());
```

Comment:
`generatePartitions(sources.size()).forEach(partition -> {...})`

Reply:
It's admittedly more fluent, but terribly painful if you need to debug something.

Reply:
Since we return `Stream` i assumed that this is for memory reasons. We rather wouldn't want to materialize all partition combinations. If we don't care about extra memory, why don't just return the materialized sets from the `generatePartitions` method?

Reply:
I agree.  The current implementation must be leftover from some time when the stream was directly used.  I'll change it to return the materialized collection since I think that's simpler and apparently we haven't encountered any memory problems.